mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -734,12 +703,10 @@ private[ml] object RandomForest extends Logging {
               val rightChildStats =
                 binAggregates.getImpurityCalculator(nodeFeatureOffset, numSplits)
               rightChildStats.subtract(leftChildStats)
-              predictionAndImpurity = Some(predictionAndImpurity.getOrElse(
-                calculatePredictImpurity(leftChildStats, rightChildStats)))
-              val gainStats = calculateGainForSplit(leftChildStats,
-                rightChildStats, binAggregates.metadata, predictionAndImpurity.get._2)
-              (splitIdx, gainStats)
-            }.maxBy(_._2.gain)
+              gainAndImpurityStats = Some(calculateImpurityStats(
```

Comment:
Here we don't need to use ```getOrElse``` because the ```gainAndImpurityStats``` need to be recompute for each split candidate, the only changeless members are ```impurity``` and ```impurityCalculator```. I have make optimization in the function calculateImpurityStats to directly get these two variables rather than recompute them.